docs/custom_vjp_update.md

```diff
@@ -0,0 +1,129 @@
+# `custom_vjp` and `nondiff_argnums` update guide
+_mattjj@_
+_Oct 14 2020_
+
+This doc assumes familiarity with `jax.custom_vjp`, as described in the [Custom
+derivative rules for JAX-transformable Python
+functions](https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html)
+notebook.
+
+### What to update
+
+After JAX [PR #4008](https://github.com/google/jax/pull/4008), the arguments
+passed into a `custom_vjp` function's `nondiff_argnums` can't be `Tracer`s (or
+containers of `Tracer`s), which basically means to allow for
+arbitrarily-transformable code `nondiff_argnums` shouldn't be used for
+array-valued arguments. Instead, `nondiff_argnums` should be used only for
+non-array values, like Python callables or shape tuples or strings.
+
+Wherever we used to use `nondiff_argnums` for array values, we should just pass
+those as regular arguments. In the `bwd` rule, we need to produce values for them,
+but we can just produce `None` values to indicate there's no corresponding
+gradient value.
+
+For example, here's the **old** way to write `clip_gradient`, which won't work
+when `hi` and/or `lo` are `Tracer`s from some JAX transformation.
+
+```python
+from functools import partial
+import jax
+
+@partial(jax.custom_vjp, nondiff_argnums=(0, 1))
+def clip_gradient(lo, hi, x):
+  return x  # identity function
+
+def clip_gradient_fwd(lo, hi, x):
+  return x, None  # no residual values to save
+
+def clip_gradient_bwd(lo, hi, _, g):
+  return (jnp.clip(g, lo, hi),)
+
+clip_gradient.defvjp(clip_gradient_fwd, clip_gradient_bwd)
+```
+
+Here's the **new**, awesome way, which supports arbitrary transformations:
+
+```python
+import jax
+
+@jax.custom_vjp  # no nondiff_argnums!
+def clip_gradient(lo, hi, x):
+  return x  # identity function
+
+def clip_gradient_fwd(lo, hi, x):
+  return x, (lo, hi)  # save lo and hi values as residuals
+
+def clip_gradient_bwd(res, g):
+  lo, hi = res
+  return (None, None, jnp.clip(g, lo, hi))  # return None for lo and hi
+
+clip_gradient.defvjp(clip_gradient_fwd, clip_gradient_bwd)
+```
+
+If you use the old way instead of the new way, you'll get a loud error in any
+case where something might go wrong (namely when there's a `Tracer` passed into
+a `nondiff_argnums` argument).
+
+Here's a case where we actually need `nondiff_argnums` with `custom_vjp`:
+
+```python
+from functools import partial
+import jax
+
+@partial(jax.custom_vjp, nondiff_argnums)
```

Comment:
This is missing the value of `nondiff_argnums`?

Reply:
Thanks! I will fix.